Geochemistry equilibrium constant interpolators

cpgr opened the pull request:
Utility to fit equilibrium constant data in the geochemistry database.

It basically constructs a fit using either a fourth-order polynomial (like GWB), or a fit used in the EQ3/6 database (and by pflotran and toughreact). 

Note that values in the database of 500 mean that there is no data, so in some cases there are not enough points to fit either of these functions. In these cases, a linear fit is used (as far as I can tell this is the same thing that toughreact and pflotran do). Also as far as I can tell, there are no cases with only 1 valid point in the database, so I didn't bother with a constant fit.

FYI @WilkAndy 

Once #14978 is complete I'll make a small change to the python converter and database to add the fit type. Then we can use these fits by constructing them once in the minimal system and then calling them like in TabulatedFluidProperties, so it should be negligible in terms of run time, even using them millions of times.

cpgr committed:
Add empty constructor in LeastSquaresFitBase

To avoid complicated matrix filling bookkeeping for the geochemistry
equilibrium constant interpolation, enable a LeastSquaresFitBase object
to be constructed without variables.

Refs #14950

cpgr committed:
Utility for fitting equilibrium constant data

A single interface to construct fits to the equilibrium constant
data in the thermodynamic database for all fits considered.

Refs #14950

cpgr commented:
Oh, also for Andy, I included an AD version in case he is thinking of using AD later on. It is a bit of copy and paste but templating it seemed overkill.

WilkAndy commented:
Thanks @cpgr :-)  I'm looking forward to exploring this next week (I'm going to have the weekend off)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/490490/) on 008d0e0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14982/site)

This comment will be updated on new commits.

WilkAndy commented on modules/geochemistry/include/utils/EquilibriumConstantInterpolator.h:
```diff
@@ -0,0 +1,64 @@
+  virtual void fillMatrix() override;
+
+  /// Functional form of fit
+  enum class FitTypeEnum
```
Can you use `const enum class` instead of just `enum class`?

WilkAndy commented on modules/geochemistry/src/utils/EquilibriumConstantInterpolator.C:
```diff
@@ -0,0 +1,171 @@
+  // A value of 500.0 in logk signifies 'no value', so should not be used in the fit
+  std::vector<Real> useful_temperature, useful_logk;
+  for (unsigned int i = 0; i < temperature.size(); ++i)
+    if (!MooseUtils::absoluteFuzzyEqual(logk[i], 500.0))
```
Can the constructor contain this value of "no value" instead of hard-coding the 500.0? (Maybe default to 500.0.)  Or is 500.0 a universally-accepted geochemistry no-value?

cpgr replied:
Good idea, I'll do this

WilkAndy replied:
Sorry to be so annoying, @cpgr, but can this 500.0000 thing somehow be in the Reader too?

WilkAndy commented on modules/geochemistry/include/utils/EquilibriumConstantInterpolator.h:
```diff
@@ -0,0 +1,64 @@
+ *
+ * a Maier-Kelly type fit (for EQ3/6 databases)
+ *
+ * log(K)= a_0 ln(T) + a_1 + a_2 T + a_3 / T + a_4 / T^2
```
Just check that it's "ln" and not "log10" - geochemists love log10 !

cpgr replied:
I think its natural log but will double check!

WilkAndy reviewed (dismissed):
This is great, thank you @cpgr .

cpgr commented:
Thanks Andy - one question. What should happen for temperatures above the highest given temperature value (300C)? At the moment this would just extrapolate, should we throw a warning?

WilkAndy commented:
> Thanks Andy - one question. What should happen for temperatures above the highest given temperature value (300C)? At the moment this would just extrapolate, should we throw a warning?

It's difficult to tell what to do at present.  I'm a bit wary about throwing lots of warnings just because i imagine the Newton process potentially going into T>300degC during convergence before finally finding a solution at T<300degC.  Remember this function could be called hundreds, possibly millions of times, per node per timestep.  Perhaps for now we could just write something in the doco?  "An extrapolation is performed for T outside the range defined in the geochemical database.  Therefore, if your simulation strays outside this range, results may not be physically realistic."  or similar.

cpgr commented:
> > Thanks Andy - one question. What should happen for temperatures above the highest given temperature value (300C)? At the moment this would just extrapolate, should we throw a warning?
> 
> It's difficult to tell what to do at present. I'm a bit wary about throwing lots of warnings just because i imagine the Newton process potentially going into T>300degC during convergence before finally finding a solution at T<300degC. Remember this function could be called hundreds, possibly millions of times, per node per timestep. Perhaps for now we could just write something in the doco? "An extrapolation is performed for T outside the range defined in the geochemical database. Therefore, if your simulation strays outside this range, results may not be physically realistic." or similar.

Sounds good. I'll add some doco

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/491246/) on 9176b96 : invalidated by @WilkAndy

wierdo errors unrelated to PR.  Invalidating as it&#39;d be great to get this merged!

cpgr committed:
Add equilibrium constant model to Geochem DB

Default logK models added to python utilities.

Refs #15146

cpgr committed:
Read equilibrium constant model from DB

Update GeochemicalDatabaseReader to read and return the
model for the equilibirium constant fit.

Closes #15146

cpgr committed:
Fill missing values in database

Avoid having missing values in the database by filling
the values during the python conversion. Add a note that
missing values have been filled.

Refs #15145

cpgr committed:
Update database reader and unit tests

Refs #15145

cpgr committed:
Update geochemical database

Updated database with missing values filled according to the fit
model.

Due to fitting, string representation of reals has been changed to
float representation.

Closes #15145

cpgr commented:
Hi Andy,

Wow - now this looks like one of your PR's! But most of the 20K line changes are removing quotes in the database around floats. These were previously strings but are now floats in the json.

Now, species that had missing values have data that looks like:
```
"BaCO3": {
      "species": {
        "Ba++": 1.0,
        "HCO3-": 1.0,
        "H+": -1.0
      },
      "charge": 0.0,
      "radius": 4.0,
      "molecular weight": 197.3392,
      "logk": [
        10.6297,
        10.3512,
        10.2197,
        10.2887,
        10.5214,
        10.8711,
        11.481,
        12.7033
      ],
      "note": "Missing array values in original database have been filled using a fourth-order fit. Original values are [10.6297, 10.3512, 10.2197, 10.2887, 10.5214, 10.8711, 500.0000, 500.0000]"
    },
```

WilkAndy commented:
Yea, this looks nice and clean, thanks @cpgr.  Perhaps you were tempted to use our (already tested) C++ code for the least-squares stuff, but i'm sorta pleased you used python instead, because i think the code is clearer.

Now i can't wait for this to be merged!

WilkAndy commented:
By the way, the failures are
```
ModuleNotFoundError: No module named 'scipy'
```

cpgr committed:
Skip python unit tests if scipy isn't available

Closes #15145

cpgr commented:
> By the way, the failures are
> 
> ```
> ModuleNotFoundError: No module named 'scipy'
> ```

Yep, I'll try and skip them if no `scipy`

WilkAndy commented:
Hey everyone involved - i'd really like this merged soon.  It's quite hard to properly test the geochemistry module if we don't have temperature dependence.

cpgr commented:
I'd forgotten about this one. I see there is a conflict with the test db - I'll fix this up now

WilkAndy commented:
Cool, thanks Chris.  Maybe you can ask one of the moose guys for a review after you've done it.  I think they've forgotten about it too!

loganharbour commented:
What's the status of this?

cpgr commented:
If Andy is happy with the changes then it is good to be reviewed/merged.

It looks like a lot of lines of code, but there is only a small amount. Most of the changes are in the database format, where floats were strings but are now floats.

WilkAndy commented:
Yea, i'm certainly happy for this to be merged, except i'm sure it'll introduce conflicts into my latest PR!  Too bad on me.

loganharbour commented:
So... which would you prefer first?

WilkAndy commented:
Let's get this one merged first.

loganharbour reviewed (approved):
👍
